File: src/main/java/services/implementations/TestDeviceManager.java
package services.implementations;


import models.TestDevice;
import spark.utils.StringUtils;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * For managing the instance of the appium driver with Thread local
 */
public class TestDeviceManager {

    private static final String A_DEVICES = "devices";
    private static final String A_VERSION = "ro.build.version.release";
    private static final String[] A_UUID = {"get", "secure", "android_id"};
    private static final String ADB = System.getenv("ANDROID_HOME") + "/platform-tools/adb";
    /**
     * iOS Commands
     */
    private static final String[] i_DEVICES = {"instruments", "-s", A_DEVICES};
    /**
     * ADB Commands [Android]
     */
    private static PrintStream LOG = System.out;
    private static List<TestDevice> testDevices = null;

    public static List<TestDevice> getTestDevices() {
        return testDevices;
    }

    public static void setTestDevices(List<TestDevice> devices) {
        testDevices = devices;
    }

    /**
     * @return
     */
    public static List<TestDevice> getConnectedDevices() {
        List<TestDevice> testDeviceList = new ArrayList<>();
        String devices[] = null;

        LOG.print("Read ANDROID_HOME  : '" + ADB + "'.");

        if (StringUtils.isEmpty(ADB)) {
            LOG.print("Set environment variable 'ANDROID_HOME' to continue.\n Hint: if ANDROID_HOME is set, try running the tool with root privileges.");
            return testDeviceList;
        }

        // get android devices
        String androidDevices = execute(ADB, A_DEVICES);
        if (StringUtils.isEmpty(androidDevices)) {
            LOG.print("ANDROID devices not found");
        }
        devices = androidDevices.split("\\r?\\n");
        for (String device : devices) {
            LOG.print("PROCESSING - '" + device);
            String temp[] = device.split("\\s+");
            if (temp.length != 2) {
                continue;
            }
            String deviceName = temp[0];
            String platformVersion = adbShell(deviceName, "getprop", A_VERSION).replace("\n", "");
            String udid = adbShell(deviceName, "settings", A_UUID).replace("\n", "");

            testDeviceList.add(new TestDevice(deviceName, "ANDROID", platformVersion, udid));
        }

        // get ios devices
        /*
         * Alternativey, we can also use this logic to get all devices : String
         * iosDevices = UtilityHelper.execute( "idevice_id","-l" ); devices =
         * iosDevices.split( "\\r?\\n" ); for ( String device : devices ) { String
         * allInfo = UtilityHelper.execute( "ideviceinfo", "-u", device); }
         */
        String iosDevices = execute(i_DEVICES);
        if (StringUtils.isEmpty(iosDevices)) {
            System.out.println("iOS devices not found");
        }

        devices = iosDevices.split("\\r?\\n");
        for (String device : devices) {
            if (device.contains("(Simulator)")) {
                continue;
            }

            List<String> match = getRegexMatch("\\(.*\\)\\s\\[.*\\]", device);
            if (match.size() > 0) {
                String deviceID = getRegexMatch("\\[(.*)\\]", device).get(1);
                String platformVersion = getRegexMatch("\\((\\d.*)\\)", device).get(1);
                String deviceName = getRegexMatch("(.*?)\\(", device).get(1);
                TestDevice testDevice = new TestDevice(deviceName, "IOS", platformVersion, deviceID);
                System.out.println("Added device - '" + testDevice + "'");
                testDeviceList.add(testDevice);
            }
        }
        System.out.println("Found " + testDeviceList.size() + " live device(s)");
        return testDeviceList;
    }

    /**
     * Utility to get regex matches
     *
     * @param regex
     * @param input
     * @return List<String> containing all matches
     */
    public static List<String> getRegexMatch(String regex, String input) {
        List<String> matches = new ArrayList<>();

        Pattern p = Pattern.compile(regex);
        Matcher m = p.matcher(input);

        while (m.find()) {
            for (int i = 0; i <= m.groupCount(); i++) {
                matches.add(m.group(i));
            }
        }
        return matches;
    }

    /**
     * Execute the command on system and returns its output as String
     *
     * @param cmd
     * @return String, output of command if none empty string is returned
     */
    public static String execute(String... cmd) {
        System.out.println("excuting command line - '" + String.join(",", cmd) + "'");

        String NEW_LINE = System.getProperty("line.separator");
        StringBuilder response = new StringBuilder();
        try {
            ProcessBuilder processBuilder = new ProcessBuilder(cmd);
            // from the doc: initially, this property is false, meaning that the
            // standard output and error output of a subprocess are sent to two
            // separate streams
            processBuilder.redirectErrorStream(true);
            Process process = processBuilder.start();

            BufferedReader in = new BufferedReader(new InputStreamReader(process.getInputStream()));
            String line;
            while ((line = in.readLine()) != null) {
                response.append(line).append(NEW_LINE);
            }
            process.waitFor();
            in.close();
        } catch (IOException | InterruptedException ex) {
            response.append(ex.getMessage());
        }
        String res = response.toString();
        LOG.print("Command line response - " + res);
        return res;
    }

    public static String adbShell(
            String deviceName, String action, String... shellCommand) {
        List<String> commands = new ArrayList<>();

        commands.add(ADB);
        commands.add("-s");
        commands.add(deviceName);
        commands.add("shell");

        switch (action) {
            case "action":
                commands.add("am");
                commands.add("start");
                commands.add("-a");
                break;

            case "getprop":
                commands.add("getprop");
                break;

            case "settings":
                commands.add("settings");
                break;

            default:
                LOG.print("Shell action '" + action + "' is not yet supported.");
                return "";
        }

        for (String s : shellCommand) {
            commands.add(s);
        }
        LOG.print("Shell command - " + commands);
        return execute(commands.toArray(new String[0]));
    }
}

File: src/main/kotlin/services/implementations/TestDeviceManager.kt
package services.implementations


import com.google.gson.Gson
import com.google.gson.JsonArray
import com.google.gson.JsonObject

import models.TestDevice
import org.apache.commons.lang3.SystemUtils
import spark.utils.StringUtils

/**
 * For managing the instance of the appium driver with Thread local
 */
object TestDeviceManager {
    private val logger = System.out
    val adb = System.getenv("ANDROID_HOME") + "/platform-tools/adb"


    /**
     * Fetch live ios device connected to your machine and return the listOf<TestDevice>
     */
    fun fetchLiveIosDevices(): List<TestDevice> {
        if (SystemUtils.IS_OS_WINDOWS) return emptyList() // windows machine

        // fetch active simulator's
        var listOfiOSDevices: List<TestDevice> = emptyList()
        var iosDevices = Gson().fromJson(Utility.execute("xcrun", "simctl", "list", "--json"), JsonObject::class.java)
        iosDevices = iosDevices["devices"] as JsonObject
        iosDevices.entrySet().forEach { entry ->
            if (entry.key.contains("iOS")) {
                (iosDevices[entry.key] as JsonArray).forEach {
                    val d = it as JsonObject
                    if (d["state"].asString == "Booted")
                        listOfiOSDevices = listOfiOSDevices.plus(
                                TestDevice(
                                        deviceName = d["name"].asString,
                                        platformName = "IOS",
                                        udid = d["udid"].asString,
                                        platformVersion = "\\d.*".toRegex().find(entry.key)!!.groups[0]!!.value.replace("-", ".")))
                }
            }
        }

        // fetch active real devices
        Utility.execute("instruments", "-s", "devices").split("\n").forEach {
            if (!it.contains("(Simulator)") && ".*\\(.*\\)\\s\\[.*\\]".toRegex().matches(it)) {
                listOfiOSDevices.plus(
                        TestDevice(
                                deviceName = Regex("(.*?)\\(").find(it)!!.groups[1]!!.value,
                                platformVersion = Regex("\\((.*?)\\)").find(it)!!.groups[1]!!.value,
                                udid = Regex("\\[(.*?)\\]").find(it)!!.groups[1]!!.value,
                                platformName = "IOS"
                        )
                )
            }
        }
        return listOfiOSDevices
    }

    /**
     * Fetch live android device connected to your machine and return the listOf<TestDevice>
     */
    fun fetchLiveAndroidDevices(): List<TestDevice> {
        var listOfAndroidDevices: List<TestDevice> = emptyList()
        logger.print("Read ANDROID_HOME  : $adb.")
        if (StringUtils.isEmpty(adb)) {
            logger.print("Set environment variable 'ANDROID_HOME' to continue.\n Hint: if ANDROID_HOME is set, try running the tool with root privileges.")
            return listOfAndroidDevices
        }
        val androidDevices = Utility.execute(adb, "devices").split("\n")
        for (device in androidDevices) {
            logger.println("PROCESSING - '$device'")
            val temp = device.split("\\s+".toRegex())
            if (temp.size != 2) continue
            listOfAndroidDevices = listOfAndroidDevices.plus(
                    TestDevice(
                            deviceName = temp[0],
                            platformVersion = Utility.adbShell(temp[0], ADBShellAction.PROPERTY, "ro.build.version.release"),
                            udid = temp[0],
                            platformName = "ANDROID"
                    )
            )
        }
        return listOfAndroidDevices;
    }
}